src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,722 @@
+        resolve_endpoints(netdev);
+
+        return 0;
+}
+
+static int wireguard_resolve_handler(sd_resolve_query *q,
+                                     int ret,
+                                     const struct addrinfo *ai,
+                                     void *userdata) {
+        NetDev *netdev;
+        Wireguard *w;
+        _cleanup_(endpoint_freep) WireguardEndpoint *e;
+        int r;
+
+        assert(userdata);
+        e = userdata;
+        netdev = e->netdev;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_REMOVE(endpoints, w->unresolved_endpoints, e);
+        w->n_unresolved_endpoints--;
+
+        if (ret != 0) {
+                log_netdev_error(netdev, "Failed to resolve host '%s:%s': %s", e->host, e->port, gai_strerror(ret));
+                LIST_PREPEND(endpoints, w->failed_endpoints, e);
+                w->n_failed_endpoints++;
+                e = NULL;
+        } else if ((ai->ai_family == AF_INET && ai->ai_addrlen == sizeof(struct sockaddr_in)) ||
+                        (ai->ai_family == AF_INET6 && ai->ai_addrlen == sizeof(struct sockaddr_in6)))
+                memcpy(&e->peer->endpoint, ai->ai_addr, ai->ai_addrlen);
+        else
+                log_netdev_error(netdev, "Neither IPv4 nor IPv6 address found for peer endpoint: %s:%s", e->host, e->port);
+
+        if (w->n_unresolved_endpoints && w->n_unresolved_endpoints % MAX_CONCURRENT_QUERIES == 0)
+                resolve_endpoints(netdev);
+        else if (!w->n_unresolved_endpoints) {
+                set_wireguard_interface(netdev);
+                if (w->failed_endpoints) {
+                        w->retries++;
+                        r = sd_event_add_time(netdev->manager->event,
+                                              &w->resolve_retry_event_source,
+                                              CLOCK_MONOTONIC,
+                                              /* cap at ~25s */
+                                              now(CLOCK_MONOTONIC) + (2 << MAX(w->retries, (unsigned)7)) * 100 * USEC_PER_MSEC,
+                                              0,
+                                              on_resolve_retry,
+                                              netdev);
+                        if (r < 0)
+                                log_netdev_warning_errno(netdev, r, "Could not arm resolve retry handler: %m");
+                }
+        }
+
+        return 0;
+}
+
+static void resolve_endpoints(NetDev *netdev) {
+        int r, i = 0;
+        Wireguard *w;
+        WireguardEndpoint *endpoint;
+        static const struct addrinfo hints = {
+                .ai_family = AF_UNSPEC,
+                .ai_socktype = SOCK_DGRAM,
+                .ai_protocol = IPPROTO_UDP
+        };
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_FOREACH(endpoints, endpoint, w->unresolved_endpoints) {
+                if (++i >= MAX_CONCURRENT_QUERIES)
```

Comment:
chunking concurrent queries is broken, as `wireguard_done` is called when this function returns..